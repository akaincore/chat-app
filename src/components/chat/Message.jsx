import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import classNames from 'classnames';
import ChatAvatar from '../common/ChatAvatar';
import moment from 'moment';

const styles = theme => ({
  root: {
    ...theme.mixins.gutters(),
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2,
    maxWidth: '70%',
  },
  messageContainer: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start',
    margin: '20px 0',
  },
  avatar: {
    margin: '0 20px',
  },
  statusMessage: {
    width: '100%',
    textAlign: 'center',
    fontStyle: 'italic',
    color: '#bbb',
  },
  statusMessageTime: {
    width: '100%',
    textAlign: 'center',
    fontStyle: 'italic',
    color: '#bbb',
    fontSize: '10px',
  },
  myMessage: {
    flexDirection: 'row-reverse',
  },
  myMessageBackground: {
    backgroundColor: '#66ccff',
  },
  emptyMessage: {
    fontStyle: 'italic',
    color: '#aaa',
  },
});

const Message = ({ classes, sender, content, currentUser, statusMessage, createdAt }) => {

  if (statusMessage) {
    return (
      <div className={classNames(classes.messageContainer)}>
        <Typography component="div" className={classes.statusMessage}>
          <Typography component="p">
            {sender.username} {content}
          </Typography>
          <Typography component="p" className={classes.statusMessageTime}>
            {moment(createdAt).fromNow()}
          </Typography>
        </Typography>
      </div>
    );
  }

  const isMy = sender._id === currentUser._id;
  return (
    <div
      className={classNames(classes.messageContainer, isMy && classes.myMessage)}
    >
      <ChatAvatar
        className={classes.avatar}
        title={sender.username}
        id={sender._id}
      />
      <Paper
        className={classNames(classes.root, isMy && classes.myMessageBackground)}
        elevation={1}
      >
        <Typography variant="subtitle2" component="h3">
          {sender.username}
        </Typography>
        <Typography component="p" className={!content ? classes.emptyMessage : ''}>
          {content || 'Message is empty'}
        </Typography>
        <Typography variant="caption" component="span">
          {moment(createdAt).fromNow()}
        </Typography>
      </Paper>
    </div>
  );
};

export default withStyles(styles)(Message);
