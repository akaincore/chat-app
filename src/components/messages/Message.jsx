import React from 'react';
import Avatar from '@material-ui/core/Avatar';
import {withStyles} from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import initials from "../../utils/initials";
import classnames from 'classnames';

const styles = theme => ({
  root: {
    ...theme.mixins.gutters(),
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2,
    maxWidth: '70%',
  },
  messageContainer: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start',
    margin: '20px 0',
  },
  avatar: {
    margin: '0 20px',
  },
  myMessage: {
    flexDirection: 'row-reverse',
  },
  myMessageBackground: {
    backgroundColor: '#66ccff',
  }
});

const Message = props => {
  const {classes} = props;
  const isMy = props.sender === 'me';
  return (
    <div
      className={classnames(classes.messageContainer, isMy && classes.myMessage)}
    >
      <Avatar className={classes.avatar}>
        {initials(props.sender)}
      </Avatar>
      <Paper
        className={classnames(classes.root, isMy && classes.myMessageBackground)}
        elevation={1}
      >
        <Typography variant="subtitle2" component="h3">
          {props.sender}
        </Typography>
        <Typography component="p">
          {props.content}
        </Typography>
      </Paper>
    </div>
  );
};

export default withStyles(styles)(Message);
