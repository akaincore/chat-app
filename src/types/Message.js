import PropTypes from 'prop-types';
import User from './User';

export default {
  chatId: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  sender: PropTypes.shape(User).isRequired,
  createdAt: PropTypes.string.isRequired,
};
