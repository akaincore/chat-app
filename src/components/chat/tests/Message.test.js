/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Message from '../Message';

jest.mock('../../common/ChatAvatar', () => () => 'ChatAvatar');

const mockProps = {
  sender: {
    _id: '12345',
    username: 'me',
  },
  content: '',
  currentUser: {
    username: 'user',
    isMember: false,
    isChatMember: false,
    isCreator: false,
  },
  statusMessage: false,
  createdAt: '2019-01-01T00:00:00.000Z',
};

describe('<Message />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Message {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Message {...mockProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
