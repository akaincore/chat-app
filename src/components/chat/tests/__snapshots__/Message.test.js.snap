// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message /> renders correctly 1`] = `
<div
  className="Message-messageContainer-73"
>
  ChatAvatar
  <div
    className="MuiPaper-root-80 MuiPaper-elevation1-83 MuiPaper-rounded-81 Message-root-72"
  >
    <h3
      className="MuiTypography-root-107 MuiTypography-subtitle2-126"
    >
      me
    </h3>
    <p
      className="MuiTypography-root-107 MuiTypography-body1-116 Message-emptyMessage-79"
    >
      Message is empty
    </p>
    <span
      className="MuiTypography-root-107 MuiTypography-caption-117"
    >
      11 days ago
    </span>
  </div>
</div>
`;
